engine/src/main/java/org/kigalisim/engine/support/SalesRecalcStrategy.java

```diff
@@ -0,0 +1,345 @@
+
+    EngineNumber manufactureSalesConverted = unitConverter.convert(manufactureRaw, "kg");
+    EngineNumber importSalesConverted = unitConverter.convert(importRaw, "kg");
+    EngineNumber priorRecycleSalesConverted = unitConverter.convert(priorRecycleRaw, "kg");
+
+    BigDecimal manufactureSalesKg = manufactureSalesConverted.getValue();
+    BigDecimal importSalesKg = importSalesConverted.getValue();
+    BigDecimal priorRecycleSalesKg = priorRecycleSalesConverted.getValue();
+    BigDecimal totalNonRecycleKg = manufactureSalesKg.add(importSalesKg);
+
+    // Get stream percentages for allocation
+    BigDecimal percentManufacture;
+    BigDecimal percentImport;
+
+    if (totalNonRecycleKg.compareTo(BigDecimal.ZERO) == 0) {
+      EngineNumber manufactureInitialCharge = engine.getInitialCharge("manufacture");
+      EngineNumber importInitialCharge = engine.getInitialCharge("import");
+      BigDecimal manufactureInitialChargeVal = manufactureInitialCharge.getValue();
+      BigDecimal importInitialChargeVal = unitConverter
+          .convert(importInitialCharge, manufactureInitialCharge.getUnits()).getValue();
+      BigDecimal totalInitialChargeVal = manufactureInitialChargeVal.add(importInitialChargeVal);
+
+      if (totalInitialChargeVal.compareTo(BigDecimal.ZERO) == 0) {
+        percentManufacture = BigDecimal.ONE;
+        percentImport = BigDecimal.ZERO;
+      } else {
+        percentManufacture = divideWithZero(manufactureInitialChargeVal, totalInitialChargeVal);
+        percentImport = divideWithZero(importInitialChargeVal, totalInitialChargeVal);
+      }
+    } else {
+      percentManufacture = divideWithZero(manufactureSalesKg, totalNonRecycleKg);
+      percentImport = divideWithZero(importSalesKg, totalNonRecycleKg);
+    }
+
+    // Recycle
+    EngineNumber newRecycleValue = new EngineNumber(recycledDisplacedKg, "kg");
+    streamKeeper.setStream(application, substance, "recycle", newRecycleValue);
+
+    // New values
+    BigDecimal requiredKgUnbound = kgForRecharge.add(kgForNew);
+    boolean requiredKgNegative = requiredKgUnbound.compareTo(BigDecimal.ZERO) < 0;
+    BigDecimal requiredKg = requiredKgNegative ? BigDecimal.ZERO : requiredKgUnbound;
+    BigDecimal newManufactureKg = percentManufacture.multiply(requiredKg);
+    BigDecimal newImportKg = percentImport.multiply(requiredKg);
+    EngineNumber newManufacture = new EngineNumber(newManufactureKg, "kg");
+    EngineNumber newImport = new EngineNumber(newImportKg, "kg");
+    streamKeeper.setStream(application, substance, "manufacture", newManufacture);
+    streamKeeper.setStream(application, substance, "import", newImport);
+  }
+
+  @Override
+  public void execute(Engine target, RecalcKit kit) {
+    // Move the logic from SingleThreadEngine.recalcSales
+    Scope scopeEffective = scope != null ? scope : target.getScope();
+
+    OverridingConverterStateGetter stateGetter =
+        new OverridingConverterStateGetter(kit.getStateGetter().orElseThrow(
```

Comment:
I think this is frequently repeated. Can we make getStateGetter return a realized StateGetter and have the RecalcKitBuilder throw an exception if one is not given?